src/ast/sym.cc

```diff
@@ -247,27 +296,389 @@ namespace sym
+  {
+    if (ast && (ast->tag == "typeref"_))
+      prec_staticcall(ast, err);
+  }
+
+  void prec_member(ast::Ast& ast, err::Errors& err)
+  {
+    prec_obj(ast, err);
+    auto next = ast::get_next_in_expr(ast);
+
+    while (next)
+    {
+      switch (next->tag)
+      {
+        case "lookup"_:
+        {
+          // (member obj lookup)
+          auto obj = ast;
+          auto member = ast::node(next, "member");
+          ast::replace(ast, member);
+          ast::push_back(ast, obj);
+          ast::remove(next);
+          ast::push_back(ast, next);
+          next = ast::get_next_in_expr(ast);
+          break;
+        }
+
+        case "typeargs"_:
+        {
+          // (invoke obj typeargs args)
+          auto obj = ast;
+          auto invoke = ast::node(next, "invoke");
+          ast::replace(ast, invoke);
+          ast::push_back(ast, obj);
+          ast::remove(next);
+          ast::push_back(ast, next);
+          next = ast::get_next_in_expr(ast);
+
+          if (next && (next->tag == "tuple"_))
+          {
+            ast::remove(next);
+            ast::rename(next, "args");
+            ast::push_back(ast, next);
+          }
+          else
+          {
+            // zero argument invoke
+            auto args = ast::node(ast, "args");
+            ast::push_back(ast, args);
+          }
+          break;
+        }
+
+        case "tuple"_:
+        {
+          // (invoke obj typeargs args)
+          auto obj = ast;
+          auto invoke = ast::node(next, "invoke");
+          ast::replace(ast, invoke);
+          ast::push_back(ast, obj);
+          auto typeargs = ast::node(ast, "typeargs");
+          ast::push_back(ast, typeargs);
+          ast::remove(next);
+          ast::rename(next, "args");
+          ast::push_back(ast, next);
+          break;
+        }
+
+        default:
+          return;
+      }
+    }
+  }
+
+  void prec_prefix(ast::Ast& ast, err::Errors& err)
+  {
+    // (call function typeargs obj (args))
+    if (ast && (ast->tag == "op"_))
+    {
+      auto op = ast;
+      auto call = ast::node(ast, "call");
+      ast::replace(ast, call);
+      ast::rename(op, "function");
+      ast::push_back(ast, op);
+      auto next = ast::get_next_in_expr(ast);
+      add_typeargs(ast, next);
+      prec_prefix(next, err);
+
+      if (!next)
+      {
+        err << ast << "Expected an argument to this prefix function."
+            << err::end;
+        return;
+      }
+
+      ast::remove(next);
+      ast::push_back(ast, next);
+      auto args = ast::node(ast, "args");
+      ast::push_back(ast, args);
+    }
+    else
+    {
+      prec_member(ast, err);
+    }
+  }
+
+  void prec_infix(ast::Ast& ast, err::Errors& err)
+  {
+    // (call function typeargs lhs (args rhs))
+    // Infix operators are left associative.
+    prec_prefix(ast, err);
+    auto next = ast::get_next_in_expr(ast);
+    std::string op;
+
+    while (next)
+    {
+      if (next->tag == "assign"_)
```

Comment:
This feels a little awkward.  I wonder if the peg could be redesigned a little to give you a list of assigns, then you wouldn't have to know about the outer level here?  We are treating `=` specially.

Reply:
That may be a good approach, but I don't know how I would do it right now. I'd rather get the PR in and look at that later.

Reply:
Sure